handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1447,65 +1484,224 @@ private static ByteBuffer toByteBuffer(ByteBuf out, int index, int len) {
                 out.nioBuffer(index, len);
     }
 
+    private static boolean inEventLoop(Executor executor) {
+        return executor instanceof EventExecutor && ((EventExecutor) executor).inEventLoop();
+    }
+
     /**
-     * Fetches all delegated tasks from the {@link SSLEngine} and runs them via the {@link #delegatedTaskExecutor}.
-     * If the {@link #delegatedTaskExecutor} is {@link ImmediateExecutor}, just call {@link Runnable#run()} directly
-     * instead of using {@link Executor#execute(Runnable)}.  Otherwise, run the tasks via
-     * the {@link #delegatedTaskExecutor} and wait until the tasks are finished.
+     * Will either run the delegating task directly calling {@link Runnable#run()} and return {@code true} or will
+     * offload the delegating task using {@link Executor#execute(Runnable)} and return {@code false}.
+     *
+     * If the task is offloaded it will take care to resume its work on the {@link EventExecutor} once there are no
+     * more tasks to process.
      */
-    private void runDelegatedTasks() {
-        if (delegatedTaskExecutor == ImmediateExecutor.INSTANCE) {
+    private boolean runDelegatedTasks(boolean inUnwrap) {
+        if (delegatedTaskExecutor == ImmediateExecutor.INSTANCE || inEventLoop(delegatedTaskExecutor)) {
+            // We should run the task directly in the EventExecutor thread and not offload at all.
             for (;;) {
                 Runnable task = engine.getDelegatedTask();
                 if (task == null) {
-                    break;
+                    return true;
                 }
-
                 task.run();
             }
         } else {
-            final List<Runnable> tasks = new ArrayList<Runnable>(2);
-            for (;;) {
-                final Runnable task = engine.getDelegatedTask();
-                if (task == null) {
-                    break;
+            final Runnable task = engine.getDelegatedTask();
+            // We only call the method when NEED_TASK is returned so we must have a task!
+            assert task != null;
+            executeDelegatingTask(task, inUnwrap);
+            return false;
+        }
+    }
+
+    private void executeDelegatingTask(Runnable task, boolean inUnwrap) {
+        processTask = true;
+        try {
+            delegatedTaskExecutor.execute(new SslTask(task, inUnwrap));
+        } catch (RejectedExecutionException e) {
+            processTask = false;
```

Comment:
Any risks for some other exception occurring and processTask falsely staying true?

Reply:
I dont think so...